packages/integrations/markdoc/components/astroNode.ts

```diff
@@ -0,0 +1,78 @@
+import type { AstroInstance } from 'astro';
+import type { RenderableTreeNode } from '@markdoc/markdoc';
+import Markdoc from '@markdoc/markdoc';
+import z from 'zod';
+
+export type AstroNode =
+	| string
+	| {
+			component: AstroInstance['default'];
+			props: Record<string, any>;
+			children: AstroNode[];
+	  }
+	| {
+			tag: string;
+			attributes: Record<string, any>;
+			children: AstroNode[];
+	  };
+
+export function createAstroNode(
+	node: RenderableTreeNode,
+	components: Record<string, AstroInstance['default']> = {}
+): AstroNode {
+	components = validateComponents(components);
+
+	if (typeof node === 'string' || typeof node === 'number') {
+		return String(node);
+	} else if (node === null || typeof node !== 'object' || !Markdoc.Tag.isTag(node)) {
+		return '';
+	}
+
+	if (isCapitalized(node.name) && node.name in components) {
+		const component = components[node.name];
+		const props = node.attributes;
+
+		const children = node.children.map((child) => createAstroNode(child, components));
```

Comment:
Can we avoid recursion? Recursions put a lot of pressure on memory, especially on big files.

Reply:
Hm, I'm not sure if we can with this implementation. Markdoc ASTs are represented as nested trees, so we'd need to flatten the tree into a different shape. Not something I've attempted before, but curious if you have suggestions?

Reply:
I think we can skip it for now then. I thought there was already a working solution that we could use. 
Indeed traversing and rebuilding a tree without recursion is not that simple - but there are already proved solutions [out there](https://www.geeksforgeeks.org/inorder-tree-traversal-without-recursion/).